modules/300-prometheus/images/grafana-dashboard-provisioner/hooks/dashboard_provisioner.py

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Flant JSC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import os, shutil
+from shell_operator import hook
+from slugify import slugify
+
+root_path = f"/etc/grafana/dashboards/"
+
+
+def main(ctx: hook.Context):
+    known_uids = set()
+    malformed_dashboards = []
+    dashboard_dict = {}
+
+    for i in ctx.snapshots.get("dashboard_resources", []):
+        dashboard = i["filterResult"]
+        definition = json.loads(dashboard["definition"])
+        dashboard_name = dashboard.get('name', '')
+
+        title = definition.get("title")
+        if not title:
+            malformed_dashboards.append(dashboard_name)
+            continue
+
+        title = slugify(title)
+
+        if not definition.get("uid"):
+            print(f"ERROR: definition.uid is mandatory field missing in the dashboard {dashboard_name}")
+            malformed_dashboards.append(dashboard_name)
+            continue
+
+        uid = definition["uid"]
+        if uid in known_uids:
+            print(f"ERROR: a dashboard with the same uid is already exist: {uid}")
+            continue
+        known_uids.add(uid)
+
+        folder = dashboard.get("folder", "General")
+        file = f"{title}.json"
+
+        if folder not in dashboard_dict:
+            dashboard_dict[folder] = {}
+
+        dashboard_dict[folder][file] = definition
+
+    if len(malformed_dashboards) > 0:
+        print(f'WARN: Skipping malformed dashboards: {", ".join(malformed_dashboards)}')
+
+    cleanup_folder(root_path)
```

Comment:
It can be a problem because Grafana asynchronously reads the main folder. Can the folder be cleaned up at the end of execution?

Reply:
1. Create or update dashboards.
2. Save file paths of touched files.
3. Iterate over dashboards on disk and delete untouched.


```python
    touched_files = set()

    for folder, files in dashboard_dict.items():
        if folder == "General":
            # General folder can't be provisioned, see the link for more details
            # https://github.com/grafana/grafana/blob/3dde8585ff951d5e9a46cfd64d296fdab5acd9a2/docs/sources/http_api/folder.md#a-note-about-the-general-folder
            folder_path = root_path
        else:
            folder_path = os.path.join(root_path, folder)
            os.makedirs(folder_path, exist_ok=True)

        for file, definition in files.items():
            file_path = os.path.join(folder_path, file)
            with open(file_path, "w") as f:
                json.dump(definition, f)
            touched_files.add(filepath)
            
            
    for folder, _, files in os.walk(folder):
        for file in files:
            file_path = os.path.join(folder, file)
            if file_path in touched_files:
                continue 
            try:
                os.unlink(file_path)
            except Exception as e:
                print('WARN: Failed to delete %s. Reason: %s' % (file_path, e))     
```

Reply:
It is not necessary to delete folders because there is no hook in Grafana to react. Folders will still be in Grafana despite being deleted from the filesystem.

Reply:
too complicated, no?
Maybe it would be better to operate symlinks instead?